pen[iOS] Add a pop-up menu for selecting caption style profiles

jernoble opened the pull request:
#### 93de15db619ee087390119b3c9c246237ea22f56
<pre>
pen[iOS] Add a pop-up menu for selecting caption style profiles
<a href="https://bugs.webkit.org/show_bug.cgi?id=301376">https://bugs.webkit.org/show_bug.cgi?id=301376</a>
<a href="https://rdar.apple.com/problem/163297495">rdar://problem/163297495</a>

Reviewed by Eric Carlson.

In 301952@main, a pop-up menu was added for macOS; add a similar pop-up menu
for iOS that allows the user to pick between their preferred subtitle styles.

Test: Tools/TestWebKitAPI/Tests/WebCore/cocoa/CaptionPreferencesTests.mm

* Source/WebCore/en.lproj/Localizable.strings:
* Source/WebKit/UIProcess/_WKCaptionStyleMenuControllerIOS.mm: Added.
(-[WKCaptionStyleMenuController init]):
(-[WKCaptionStyleMenuController rebuildMenu]):
(-[WKCaptionStyleMenuController isAncestorOf:]):
(-[WKCaptionStyleMenuController captionStyleMenu]):
(-[WKCaptionStyleMenuController contextMenuInteraction]):
(-[WKCaptionStyleMenuController profileActionSelected:]):
(-[WKCaptionStyleMenuController systemCaptionStyleSettingsActionSelected:]):
(-[WKCaptionStyleMenuController notifyMenuWillOpen]):
(-[WKCaptionStyleMenuController notifyMenuDidClose]):
(-[WKCaptionStyleMenuController contextMenuInteraction:configurationForMenuAtLocation:]):
(-[WKCaptionStyleMenuController contextMenuInteraction:willDisplayMenuForConfiguration:animator:]):
(-[WKCaptionStyleMenuController contextMenuInteraction:willEndForConfiguration:animator:]):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebCore/cocoa/CaptionPreferencesTests.mm:
(-[CaptionPreferenceTestMenuControllerDelegate captionStyleMenuWillOpen:]):
(-[CaptionPreferenceTestMenuControllerDelegate captionStyleMenuDidClose:]):
(TestWebKitAPI::CaptionPreferenceTests::ensureController):
(TestWebKitAPI::CaptionPreferenceTests::ensureMenu):
(TestWebKitAPI::CaptionPreferenceTests::showMenuAndThen):
(TestWebKitAPI::CaptionPreferenceTests::runAndWaitForCaptionStyleMenuWillOpenCalled):
(TestWebKitAPI::CaptionPreferenceTests::runAndWaitForCaptionStyleMenuDidCloseCalled):
(TestWebKitAPI::CaptionPreferenceTests::selectProfileAtIndex):
(TestWebKitAPI::TEST_F(CaptionPreferenceTests, ShimTest)):
(TestWebKitAPI::TEST_F(CaptionPreferenceTests, FontFace)):
(TestWebKitAPI::TEST_F(CaptionPreferenceTests, FontSize)):
(TestWebKitAPI::TEST_F(CaptionPreferenceTests, Colors)):
(TestWebKitAPI::TEST_F(CaptionPreferenceTests, BorderRadius)):
(TestWebKitAPI::TEST_F(CaptionPreferenceTests, TextEdge)):
(TestWebKitAPI::TEST_F(CaptionPreferenceTests, CaptionStyleMenu)):
(TestWebKitAPI::TEST_F(CaptionPreferenceTests, CaptionStyleMenuHighlight)):
(TestWebKitAPI::TEST_F(CaptionPreferenceTests, CaptionStyleMenuSelect)):
(TestWebKitAPI::TEST_F(CaptionPreferenceTests, CaptionStyleMenuDelegate)):
(TestWebKitAPI::TEST(CaptionPreferenceTests, ShimTest)): Deleted.
(TestWebKitAPI::TEST(CaptionPreferenceTests, FontFace)): Deleted.
(TestWebKitAPI::TEST(CaptionPreferenceTests, FontSize)): Deleted.
(TestWebKitAPI::TEST(CaptionPreferenceTests, Colors)): Deleted.
(TestWebKitAPI::TEST(CaptionPreferenceTests, BorderRadius)): Deleted.
(TestWebKitAPI::TEST(CaptionPreferenceTests, TextEdge)): Deleted.
(TestWebKitAPI::TEST(CaptionPreferenceTests, CaptionStyleMenu)): Deleted.
(TestWebKitAPI::TEST(CaptionPreferenceTests, CaptionStyleMenuHighlight)): Deleted.
(TestWebKitAPI::TEST(CaptionPreferenceTests, CaptionStyleMenuSelect)): Deleted.

Canonical link: <a href="https://commits.webkit.org/302143@main">https://commits.webkit.org/302143@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5b1738f6ae67c71fd8e13c6e763a4c5183491f13

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/128129 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/413 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/38961 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/135494 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/79624 "Built successfully") | [❌ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/4da7d483-af84-4997-bb04-0b8564824eec) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/130001 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/334 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/285 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/97536 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/65434 "") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/3d5e2f56-4f1f-48ec-94a0-0c3838950981) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/131077 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/205 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/114778 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/78106 "Passed tests") | | [❌ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/3c1ecd9b-e8b1-400f-958e-d2f44c9a1a34) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/195 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/32885 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/78805 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/108571 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/33371 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/137984 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/265 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/249 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/106065 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/296 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/111121 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/105804 "Passed tests") | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/26979 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/164/builds/200 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/29675 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/52488 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20022 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/312 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/62129 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/221 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/294 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/271 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on Source/WebKit/UIProcess/ios/_WKCaptionStyleMenuControllerIOS.mm:
```diff
@@ -0,0 +1,186 @@
+{
+    RetainPtr stylesMenuTitle = WEB_UI_NSSTRING_KEY(@"Styles", @"Styles (Media Controls Menu)", @"Subtitles media controls menu title");
+
+    auto profileIDs = CaptionUserPreferencesMediaAF::platformProfileIDs();
+    if (profileIDs.isEmpty()) {
+        _menu = [UIMenu menuWithTitle:stylesMenuTitle.get() children:@[]];
+        return;
+    }
```
Don't you add the `Manage Styles` item on macOS even if there are no profile IDs?

jernoble replied:
Huh, that's a good point. I think the behavior here is correct, but I should add a test case for that and align the two behaviors.

jernoble committed:
pen[iOS] Add a pop-up menu for selecting caption style profiles
https://bugs.webkit.org/show_bug.cgi?id=301376
rdar://problem/163297495

Reviewed by Eric Carlson.

In 301952@main, a pop-up menu was added for macOS; add a similar pop-up menu
for iOS that allows the user to pick between their preferred subtitle styles.

Test: Tools/TestWebKitAPI/Tests/WebCore/cocoa/CaptionPreferencesTests.mm

* Source/WebCore/en.lproj/Localizable.strings:
* Source/WebKit/UIProcess/_WKCaptionStyleMenuControllerIOS.mm: Added.
(-[WKCaptionStyleMenuController init]):
(-[WKCaptionStyleMenuController rebuildMenu]):
(-[WKCaptionStyleMenuController isAncestorOf:]):
(-[WKCaptionStyleMenuController captionStyleMenu]):
(-[WKCaptionStyleMenuController contextMenuInteraction]):
(-[WKCaptionStyleMenuController profileActionSelected:]):
(-[WKCaptionStyleMenuController systemCaptionStyleSettingsActionSelected:]):
(-[WKCaptionStyleMenuController notifyMenuWillOpen]):
(-[WKCaptionStyleMenuController notifyMenuDidClose]):
(-[WKCaptionStyleMenuController contextMenuInteraction:configurationForMenuAtLocation:]):
(-[WKCaptionStyleMenuController contextMenuInteraction:willDisplayMenuForConfiguration:animator:]):
(-[WKCaptionStyleMenuController contextMenuInteraction:willEndForConfiguration:animator:]):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebCore/cocoa/CaptionPreferencesTests.mm:
(-[CaptionPreferenceTestMenuControllerDelegate captionStyleMenuWillOpen:]):
(-[CaptionPreferenceTestMenuControllerDelegate captionStyleMenuDidClose:]):
(TestWebKitAPI::CaptionPreferenceTests::ensureController):
(TestWebKitAPI::CaptionPreferenceTests::ensureMenu):
(TestWebKitAPI::CaptionPreferenceTests::showMenuAndThen):
(TestWebKitAPI::CaptionPreferenceTests::runAndWaitForCaptionStyleMenuWillOpenCalled):
(TestWebKitAPI::CaptionPreferenceTests::runAndWaitForCaptionStyleMenuDidCloseCalled):
(TestWebKitAPI::CaptionPreferenceTests::selectProfileAtIndex):
(TestWebKitAPI::TEST_F(CaptionPreferenceTests, ShimTest)):
(TestWebKitAPI::TEST_F(CaptionPreferenceTests, FontFace)):
(TestWebKitAPI::TEST_F(CaptionPreferenceTests, FontSize)):
(TestWebKitAPI::TEST_F(CaptionPreferenceTests, Colors)):
(TestWebKitAPI::TEST_F(CaptionPreferenceTests, BorderRadius)):
(TestWebKitAPI::TEST_F(CaptionPreferenceTests, TextEdge)):
(TestWebKitAPI::TEST_F(CaptionPreferenceTests, CaptionStyleMenu)):
(TestWebKitAPI::TEST_F(CaptionPreferenceTests, CaptionStyleMenuHighlight)):
(TestWebKitAPI::TEST_F(CaptionPreferenceTests, CaptionStyleMenuSelect)):
(TestWebKitAPI::TEST_F(CaptionPreferenceTests, CaptionStyleMenuDelegate)):
(TestWebKitAPI::TEST(CaptionPreferenceTests, ShimTest)): Deleted.
(TestWebKitAPI::TEST(CaptionPreferenceTests, FontFace)): Deleted.
(TestWebKitAPI::TEST(CaptionPreferenceTests, FontSize)): Deleted.
(TestWebKitAPI::TEST(CaptionPreferenceTests, Colors)): Deleted.
(TestWebKitAPI::TEST(CaptionPreferenceTests, BorderRadius)): Deleted.
(TestWebKitAPI::TEST(CaptionPreferenceTests, TextEdge)): Deleted.
(TestWebKitAPI::TEST(CaptionPreferenceTests, CaptionStyleMenu)): Deleted.
(TestWebKitAPI::TEST(CaptionPreferenceTests, CaptionStyleMenuHighlight)): Deleted.
(TestWebKitAPI::TEST(CaptionPreferenceTests, CaptionStyleMenuSelect)): Deleted.

Canonical link: https://commits.webkit.org/302143@main
